accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class RemovedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between and (2) are disjoint (such as iOS and Windows), the
+  information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
+2. '{API}' has been deprecated since {OS} {OS version}.
+3. '{API}' has been removed since {OS} {OS version}.
+
+### Automatic suppression via platform guards
+
+Diagnostic (1) can be suppressed via a call to
```

Comment:
Some existing code uses Environment.OSVersion.  Presumably we'd want to factor that in when possible?

Reply:
I wouldn't. In fact, we should obsolete `Environment.OSVersion`. Two reasons:

1. `Environment.OSVersion` doesn't return the correct version number today (it's shimmed). We could change but that would just further the lifetime of a broken API. Wes should push folks to `RuntimeInformation.IsOSPlatformOrLater()`
2. Make the analyzer more complex and more prone to false positives

Reply:
I have added this as a question to the Q&A section.

Reply:
> `Environment.OSVersion` doesn't return the correct version number today 

We have fixed that. It is going to return the correct version for .NET 5+ on Windows.

Reply:
> Make the analyzer more complex and more prone to false positives

I don't see how. Factoring it in will only serve to reduce possible false positives, not increase them. If you just ignore the check, what it was guarding will be seen as unguarded.

Reply:
> I don't see how. Factoring it in will only serve to reduce possible false positives, not increase them.

Assuming we get them right. That's why I'd rather scope this out and tell people to use `RuntimeInformation` instead.